### ISSUE TITLE
ERR_CONVERSATION_LIMIT

### ISSUE THREAD

**Issue by h9052300** [mrgick/duck_chat#17]
I'll always be in trouble.
{“action”: “error”, “status”:429, “type”: “ERR_CONVERSATION_LIMIT”}

**Comment by k-aito** [comment]
I think it could be a proper error display, but I don't think it can be bypassed. You can reset the chat, and sometimes it helps, but it also stops working after a time. I think it's a limit on the DDG side.

**Comment by Troptrap** [comment]
Proxies and maybe changing the user agent randomly in duck_chat could help. Somehow DDG identifies the device sending too many requests and throttles it.
For me it happens only when I abuse it too much.

**Comment by mrgick** [comment]
@h9052300 
Try to open https://duck.ai in browser - if there is error too, so you need some proxy

**Comment by k-aito** [comment]
I think it's related that when you use it too much, they just force you to reset the chat. It would not really help to use a proxy or something else, in my opinion.

**Comment by Troptrap** [comment]
> I think it's related that when you use it too much, they just force you to reset the chat. It would not really help to use a proxy or something else, in my opinion.

I'm using it programmatically, changing the model on each completion and I'm still getting throttled. I assume this is what you mean by reset

**Comment by k-aito** [comment]
Yes, when I use it for long conversation after something like 10 messages, it forces making another conversation. But it's done too much, it just blocks you for some time.

I don't know if it was also like that on the browser, but I could no more use the application for a bit of time.

**Comment by lunaticsm** [comment]
I guess that's because DDC issue, You can make each question are one session.

**Comment by Troptrap** [comment]
> I guess that's because DDC issue, You can make each question are one session.

```
async def DDG(self,model: str, prompt: str) -> str:
    ddg_chat = DuckChat(model)
    resp = await ddg_chat.ask_question(prompt)
    await ddg_chat._session.close()
    print(resp)
    return resp
```

This is how I'm using it and still get limited. This function's argument 'model' comes from another function that randomizes the models, so it also resets history from DDG. So, there is some other pattern that DDG uses to identify same user.

**Comment by Troptrap** [comment]
@mrgick it's possible that history sent from previous model to trigger some red flag when switching models? I don't know, but I assume they have some bot detection in place

**Comment by mrgick** [comment]
Hello, sorry for long answer
------
```python
import asyncio
from duck_chat import DuckChat

async def DDG(model, prompt)
    async with DuckChat(model=model) as chat:
          resp = await ddg_chat.ask_question(prompt)
          print(resp)
    return resp


async def test():
  await DDG("gpt-4o-mini","2+2?")
  await DDG("gpt-4o-mini","Tell answer")

asyncio.run(test())
```
--------
About bot detection, I think duck.ai start to detect aiohttp connection (maybe bad cookies/tcp timings) or maybe they see that we don't send country.json and status requests.
--------
I think library need to migrate from aiohttp to curl_cffi with more simulating request (maybe this will help)

**Comment by k-aito** [comment]
I think I only got 1 time an issue. Even now, I didn't get the ERR_CONVERSATION_LIMIT yet, but it's true I only use it for a few questions, and then I close the session.

That's why I guess it's not really the library, but just the limitation that DDG choose to apply.

**Comment by Troptrap** [comment]
I have abused a little bit Duckduckgo's servers lately. I am getting a different error, not the CONVERSATION LIMIT, but ERR_UPSTREAM. I am quite confident it's an IP issue, because as soon as I switch network, it's working again. 
It could be also fingerprints, but I think bot fingerprinting would persist after IP changing.
I will be trying with proxies ASAP and will report here.

**Comment by Dalveral** [comment]
> > Думаю, это из-за проблемы DDC. Вы можете сделать каждый вопрос одной сессией.
> 
> ```
> async def DDG(self,model: str, prompt: str) -> str:
>     ddg_chat = DuckChat(model)
>     resp = await ddg_chat.ask_question(prompt)
>     await ddg_chat._session.close()
>     print(resp)
>     return resp
> ```
> 
> Вот как я его использую и все еще ограничиваюсь. Аргумент этой функции «модель» происходит от другой функции, которая рандомизирует модели, поэтому она также сбрасывает историю из DDG. Таким образом, есть еще одна модель, которую DDG использует для идентификации одного и того же пользователя.



> > I guess that's because DDC issue, You can make each question are one session.
> 
> ```
> async def DDG(self,model: str, prompt: str) -> str:
>     ddg_chat = DuckChat(model)
>     resp = await ddg_chat.ask_question(prompt)
>     await ddg_chat._session.close()
>     print(resp)
>     return resp
> ```
> 
> This is how I'm using it and still get limited. This function's argument 'model' comes from another function that randomizes the models, so it also resets history from DDG. So, there is some other pattern that DDG uses to identify same user.

Thank you. This is work!

**Comment by Dalveral** [comment]
Hello. I am getting a 429 error again. However, "await ddg_chat._session.close()" is not working now.

**Comment by Dalveral** [comment]
I used the method text.replace('\n', '') before sending requests; there were probably too many line breaks in my text. Now there are no errors. I didn't think to check this earlier because I was receiving a 429 error and thought the problem was that the service had restricted access from my device.

Thank you for your work!
